### PR TITLE
Enriching the guidance for APIs needing asynchronous support for long-running requests

### PR DESCRIPTION
**Pull request by udidahan**

### CODE REVIEW

**Comment by Court72** [comment]
@EdPrice-MSFT 

Can you review the proposed changes? *IMPORTANT*: When the changes are ready for publication, add a `#sign-off` comment to signal that the PR is ready for the review team to merge. 

#label:"aq-pr-triaged"
@MicrosoftDocs/public-repo-pr-review-team

**Review by EdPrice-MSFT — APPROVED** [review]
Looks good!

**Comment by EdPrice-MSFT** [comment]
#Sign-Off

"architectures" is a valid alias.

File: docs/best-practices/api-implementation.md [hunk 1]
```diff
@@ -634,12 +634,15 @@ The web API should also provide a mechanism to return the results of the process
...
+6. When the task has completed, the web API stores the results in the table, and it sets the state of the task to *Complete*. Note that if the task fails, the web API could also store information about the failure and set the status to *Failed*.
+  - Consider applying [retry techniques](/architecture/patterns/retry) to resolve possibly transient failures.
+8. While the task is running, the client can continue performing its own processing. It can periodically send a request to the URI it received earlier.
+9. The web API at the URI queries the state of the corresponding task in the table and returns a response message with HTTP status code 200 (OK) containing this state (*Running*, *Complete*, or *Failed*). If the task has completed or failed, the response message can also include the results of the processing or any information available about the reason for the failure.
+  - If the long-running process has more intermediate states, it's better to use a library that supports the saga pattern, like [NServiceBus](https://docs.particular.net/nservicebus/sagas) or [MassTransit](https://masstransit-project.com/usage/sagas).
```
Comment (Court72) [inline docs/best-practices/api-implementation.md:645]:
```suggestion
   - If the long-running process has more intermediate states, it's better to use a library that supports the saga pattern, like [NServiceBus](https://docs.particular.net/nservicebus/sagas) or [MassTransit](https://masstransit-project.com/usage/sagas).
```

File: docs/best-practices/api-implementation.md [hunk 2]
```diff
@@ -634,12 +634,15 @@ The web API should also provide a mechanism to return the results of the process
...
+3. The web API initiates the processing as a [separate task](/dotnet/csharp/programming-guide/concepts/async/task-asynchronous-programming-model) or with a library like [Hangfire](https://www.hangfire.io). The web API records the state of the task in the table as *Running*.
+  - If you use Azure Service Bus, the message processing would be done separately from the API, possibly by using [Azure Functions](/azure/azure-functions) or [AKS](/azure/aks).
+5. The web API returns a response message with HTTP status code 202 (Accepted), and a URI containing the unique key generated - something like */polling/{guid}*.
+6. When the task has completed, the web API stores the results in the table, and it sets the state of the task to *Complete*. Note that if the task fails, the web API could also store information about the failure and set the status to *Failed*.
+  - Consider applying [retry techniques](/architecture/patterns/retry) to resolve possibly transient failures.
```
Comment (Court72) [inline docs/best-practices/api-implementation.md:642]:
```suggestion
   - Consider applying [retry techniques](/architecture/patterns/retry) to resolve possibly transient failures.
```

File: docs/best-practices/api-implementation.md [hunk 3]
```diff
@@ -634,12 +634,15 @@ The web API should also provide a mechanism to return the results of the process
...
+2. The web API stores information about the request in a table held in [Azure Table Storage](/azure/storage/tables) or [Microsoft Azure Cache](/azure/azure-cache-for-redis), and generates a unique key for this entry, possibly in the form of a GUID. Alternatively, a message containing information about the request and the unique key could be sent via [Azure Service Bus](/azure/service-bus-messaging) as well.
+3. The web API initiates the processing as a [separate task](/dotnet/csharp/programming-guide/concepts/async/task-asynchronous-programming-model) or with a library like [Hangfire](https://www.hangfire.io). The web API records the state of the task in the table as *Running*.
+  - If you use Azure Service Bus, the message processing would be done separately from the API, possibly by using [Azure Functions](/azure/azure-functions) or [AKS](/azure/aks).
```
Comment (Court72) [inline docs/best-practices/api-implementation.md:639]:
```suggestion
   - If you use Azure Service Bus, the message processing would be done separately from the API, possibly by using [Azure Functions](/azure/azure-functions) or [AKS](/azure/aks).
```

File: docs/best-practices/api-implementation.md [hunk 4]
```diff
@@ -634,12 +634,15 @@ The web API should also provide a mechanism to return the results of the process
...
-6. While the task is running, the client can continue performing its own processing. It can periodically send a request to the URI */polling/{guid}* where *{guid}* is the GUID returned in the 202 response message by the web API.
-7. The web API at the */polling/{guid}* URI queries the state of the corresponding task in the table and returns a response message with HTTP status code 200 (OK) containing this state (*Running*, *Complete*, or *Failed*). If the task has completed or failed, the response message can also include the results of the processing or any information available about the reason for the failure.
+2. The web API stores information about the request in a table held in [Azure Table Storage](/azure/storage/tables) or [Microsoft Azure Cache](/azure/azure-cache-for-redis), and generates a unique key for this entry, possibly in the form of a GUID. Alternatively, a message containing information about the request and the unique key could be sent via [Azure Service Bus](/azure/service-bus-messaging) as well.
```
Comment (Court72) [inline docs/best-practices/api-implementation.md:637]:
```suggestion
2. The web API stores information about the request in a table held in [Azure Table Storage](/azure/storage/tables) or [Microsoft Azure Cache](/azure/azure-cache-for-redis) and generates a unique key for this entry, possibly in the form of a GUID. Alternatively, a message containing information about the request and the unique key could be sent via [Azure Service Bus](/azure/service-bus-messaging) as well.
```

**Review by Court72 — APPROVED** [review]

**Pull request merged**